src/main/kotlin/org/rust/cargo/toolchain/tools/Rustfmt.kt

```diff
@@ -29,49 +36,81 @@ fun RsToolchainBase.rustfmt(): Rustfmt = Rustfmt(this)
 class Rustfmt(toolchain: RsToolchainBase) : RustupComponent(NAME, toolchain) {
 
     fun reformatDocumentTextOrNull(cargoProject: CargoProject, document: Document): String? {
+        val project = cargoProject.project
         return createCommandLine(cargoProject, document)
-            ?.execute(cargoProject.project, stdIn = document.text.toByteArray())
-            ?.unwrapOrElse { if (isUnitTestMode) throw it else return null }
-            ?.stdout
+            ?.execute(project, stdIn = document.text.toByteArray())
+            ?.unwrapOrElse { e ->
+                e.showRustfmtError(project)
+                if (isUnitTestMode) throw e else return null
+            }?.stdout
     }
 
     fun createCommandLine(cargoProject: CargoProject, document: Document): GeneralCommandLine? {
         val file = document.virtualFile ?: return null
         if (file.isNotRustFile || !file.isValid) return null
 
-        val arguments = buildList<String> {
-            add("--emit=stdout")
+        val project = cargoProject.project
+        val settings = project.rustfmtSettings
 
-            val configPath = findConfigPathRecursively(file.parent, stopAt = cargoProject.workingDirectory)
-            if (configPath != null) {
-                add("--config-path")
-                add(configPath.toString())
-            }
+        val arguments = ParametersListUtil.parse(settings.additionalArguments).toMutableList()
 
-            val currentRustcVersion = cargoProject.rustcInfo?.version?.semver
-            if (currentRustcVersion != null) {
-                val edition = runReadAction {
-                    val psiFile = file.toPsiFile(cargoProject.project)
-                    psiFile?.edition ?: Edition.DEFAULT
-                }
-                add("--edition=${edition.presentation}")
+        val toolchain = arguments.firstOrNull()?.takeIf { it.startsWith("+") }
+        if (toolchain == null && settings.channel != RustChannel.DEFAULT) {
+            arguments.add(0, "+${settings.channel}")
+        }
+
+        val idx = arguments.indexOf("--emit")
+        if (idx in 0 until arguments.size - 1) arguments.removeAt(idx + 1)
```

Comment:
Two notes:
- it removes only the first `--emit foo` ergument
- we omit `{}` only for operations that make some "jump": `return`, `continue`, `break` and for one-line `if-else`

If I write this code, I'd write something like
```kotlin
val arguments = ParametersListUtil.parse(settings.additionalArguments)
val cleanArguments = mutableListOf<String>()

val toolchain = arguments.firstOrNull()?.takeIf { it.startsWith("+") }
when {
    toolchain != null -> cleanArguments += toolchain
    settings.channel != RustChannel.DEFAULT -> cleanArguments += "+${settings.channel}"
}

var idx = if (toolchain == null) 0 else 1
while (idx < arguments.size) {
    val arg = arguments[idx]
    when {
        arg == "--emit" -> idx += 2
        arg.startsWith("--emit") -> idx += 1
        else -> {
            cleanArguments += arg
            idx += 1
        }
    }
}
cleanArguments.add("--emit=stdout")
```

yes, it's more low-level, but it looks more clean for me